### ISSUE TITLE
Tracking Issue for BufWriter::into_raw_parts

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust#80690]
Feature gate: `#![feature(bufwriter_into_raw_parts)]`

This is a tracking issue for `BufWriter::into_raw_parts` and its associated error type etc.

This allows a `BufWriter` to be disassembled, and the inner writer to be recovered - without attempting to write out any buffered data (instead it is returnedt to the caller), and therefore (unlike `into_inner`) succeeding even if the underlying writer is returning errors.

### Public API

```rust
// std::bufwriter

/// Disassembles this `BufWriter<W>`, returning the underlying writer, and any buffered but
/// unwritten data.
impl<W> BufWriter<W> {
    pub fn into_raw_parts(mut self) -> (W, Result<Vec<u8>, WriterPanicked>);
}

/// Error newtype, wraps the buffered data, a `Vec<u8>`
pub struct WriterPanicked;
```

### Steps / History

- [ ] Implementation: #79705 
- [ ] Gain experience with this API, to see if it's good
- [ ] Final commenting period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Is there a suitable simpler API for presenting this functionality?  (I think not, see https://github.com/rust-lang/rust/pull/79705#issue-532707440)

**Comment by dtolnay** [comment]
@rust-lang/libs
@rfcbot fcp merge

Check out @ijackson's recap of this API in https://github.com/rust-lang/rust/pull/84770#issue-628289300.

**Comment by BurntSushi** [comment]
Should `BufWriter::into_raw_parts` be called `into_parts` instead? In all previous uses of "raw" in similar names, I believe a raw pointer is involved somewhere. But that isn't the case here.

Also, it looks like the `WriterPanicked` error might not actually be exported? At least, rustdoc doesn't think so: https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_raw_parts

**Comment by ijackson** [comment]
> Should `BufWriter::into_raw_parts` be called `into_parts` instead? In all previous uses of "raw" in similar names, I believe a raw pointer is involved somewhere. But that isn't the case here.

I don't have an opinion about this.  I just chose that name to distinguish it more from `into_inner`.  I would be quite happy to change it to `into_parts`.

> Also, it looks like the `WriterPanicked` error might not actually be exported? At least, rustdoc doesn't think so: https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_raw_parts

I will check this.  Thanks!

**Comment by yaahc** [comment]
> Should `BufWriter::into_raw_parts` be called `into_parts` instead? In all previous uses of "raw" in similar names, I believe a raw pointer is involved somewhere. But that isn't the case here.

I'm with @BurntSushi on this one, the `raw` in this fn name feels odd.

**Comment by ijackson** [comment]
> I'm with @BurntSushi on this one, the `raw` in this fn name feels odd.

Indeed.  I did some looking around and I think your feeling is justified:https://github.com/rust-lang/rust/pull/85901/commits/ad88a43b2fa26e35bded55a9c245c95932246e61

Hence #85901 where I change this.  That ought to go first.  I'm not sure what the implications are for the stabilisation - does that mean it should wait some more?

Thanks.

**Comment by joshtriplett** [comment]
@rfcbot concern name

Registering a concern so that this FCP doesn't complete until we finalize the name.

I agree that `into_parts` makes sense here; this is a safe method returning safe objects, unlike `Vec::into_raw_parts` which is unsafe and returns a raw pointer.

**Comment by joshtriplett** [comment]
@ijackson If the *only* thing that changes is the name, we don't need to delay stabilization (other than to wait for the PR changing the name). If something more substantive changes, we can consider if that impacts the evaluation we've done so far or not.

**Comment by m-ou-se** [comment]
The 'raw' here refers to the fact that unlike .into_inner(), this function does not attempt to flush the data but just directly gives you the 'raw' parts of the buffer, without trying to process them.

I can see how it can be confusing with our other uses of the word *raw*, but just calling it `.into_parts()` might not make clear what the difference with `.into_inner()` is. (Though on the other hand, one could argue that .into_inner has to consolidate the parts into a single 'inner' again before returning it, wheras .into_parts just returns the parts (the buffer and the underlying writer) directly.)

**Comment by ijackson** [comment]
`into_inner` gives you just an inner writer; `into_parts` gives you all the parts.  I don't have a particularly strong feeling about this but I think `into_parts` is somewhat better.  So I'll go and deal with the review comment on the fixups MR #85901.

**Comment by m-ou-se** [comment]
Looks like WriterPanicked isn't publicly available: https://doc.rust-lang.org/nightly/std/io/struct.BufWriter.html#method.into_raw_parts Did something go wrong there with the visibility?

**Comment by ijackson** [comment]
@m-ou-se: Yes, something did go wrong - I don't think I understood how all the use/visibility/stability plumbing worked.  I was trying to fix that in #85901 but there is still something wrong with that.  I will fix it tomorrow when I have a clear head.  My apologies.
